src/kakarot/stack.cairo

```diff
@@ -109,41 +191,41 @@ namespace Stack {
-        let array_index = 2 * array_index;  // convert the Uint256 array index to felt array index
-        let (dst_elts: felt*) = alloc();  // new segment for stack copy
-        let src_elts = cast(self.elements, felt*);  // convert Uint256* to felt*
-
-        memcpy(dst=dst_elts, src=src_elts, len=array_index);  // copy the stack until the swaped elt
-        assert [dst_elts + array_index] = src_elts[self.raw_len - 2];  // save the swaped elt low
-        assert [dst_elts + array_index + 1] = src_elts[self.raw_len - 1];  // save the swaped elt high
-
-        // copy the reset of the stack
-        memcpy(
-            dst=dst_elts + array_index + 2,
-            src=src_elts + array_index + 2,
-            len=self.raw_len - array_index - 4,
-        );
-
-        assert dst_elts[self.raw_len - 2] = element.low;  // copy swaped elt low to top of the stack
-        assert dst_elts[self.raw_len - 1] = element.high;  // copy swaped elt high to top of the stack
-
-        let elements = cast(dst_elts, Uint256*);  // cast dest to Uint256* to recreate a stack
-        return new model.Stack(elements=elements, raw_len=self.raw_len);
+        let stack_word_dict = self.stack_word_dict;
+        let position_zero = self.stack_16bytes_len;
+
+        // Check if there is underflow
+        Stack.check_underflow(stack_len=position_zero, pop_len=i);
+
+        // Read elements at stack postition 1
+        let (el1_high) = dict_read{dict_ptr=stack_word_dict}(position_zero - 2);
+        let (el1_low) = dict_read{dict_ptr=stack_word_dict}(position_zero - 1);
+        // Read elements at stack postition N
+        let (el2_high) = dict_read{dict_ptr=stack_word_dict}(position_zero - i * 2);
+        let (el2_low) = dict_read{dict_ptr=stack_word_dict}(position_zero - i * 2 + 1);
+
+        // Swap elements
+        dict_write{dict_ptr=stack_word_dict}(position_zero - 2, el2_high);
+        dict_write{dict_ptr=stack_word_dict}(position_zero - 1, el2_low);
+        dict_write{dict_ptr=stack_word_dict}((position_zero - i * 2), el1_high);
+        dict_write{dict_ptr=stack_word_dict}((position_zero - i * 2 + 1), el1_low);
+
+        // Return Stack
+        return (
+            new model.Stack(
+            stack_word_dict_start=self.stack_word_dict_start,
+            stack_word_dict=stack_word_dict,
+            stack_16bytes_len=self.stack_16bytes_len,
+            ));
     }
 
     // @notice Check stack overflow.
     // @param self - The pointer to the stack.
     // @custom:revert if stack overflow.
-    func check_overflow{range_check_ptr}(self: model.Stack*) {
-        let stack_len = Stack.len(self);
+    func check_overflow{range_check_ptr}(stack_len: felt) {
         // Revert if stack overflow
         with_attr error_message("Kakarot: StackOverflow") {
-            assert_lt_felt(stack_len, Constants.STACK_MAX_DEPTH);
+            assert_lt_felt(stack_len, Constants.STACK_MAX_DEPTH * 2);
```

Comment:
Use assert_le, it's less expensive

Reply:
Also, I think here it is enough to check for equality, since only push one at a time.